dynamiqs/integrators/sepropagator/expm_integrator.py

```diff
@@ -26,46 +25,33 @@ def __init__(self, *args):
             )
 
     def run(self) -> PyTree:
-        # === find times at which the Hamiltonian changes (between t0 and tfinal)
-        t0 = jnp.asarray(self.t0).reshape(-1)
-        times = _concatenate_sort(t0, self.H.discontinuity_ts, self.ts)
-        inbounds = (times[:-1] >= t0) & (times[1:] <= self.ts[-1])
-        t_diffs = jnp.where(inbounds, jnp.diff(times), 0.0)
+        # === find all times at which to stop in [t0, t1]
+        # find all times where the solution should be saved (self.ts) or at which the
+        # Hamiltonian changes (self.H.discontinuity_ts)
+        disc_ts = self.H.discontinuity_ts
+        if disc_ts is not None:
+            disc_ts = disc_ts.clip(self.t0, self.t1)
+        times = _concatenate_sort(jnp.asarray([self.t0]), self.ts, disc_ts)  # (ntimes,)
 
-        # === compute the propagators
-        # don't need the last time in times since the hamiltonian is guaranteed
-        # to be constant over the region times[-2] to times[-1]
-        Hs = jnp.stack([self.H(t) for t in times[:-1]])
-        dims = tuple(range(-Hs.ndim + 1, 0))
-        t_diffs = jnp.expand_dims(t_diffs, dims)
-        step_propagators = expm(-1j * t_diffs * Hs)
+        # === compute time differences (null for times outside [t0, t1])
+        delta_ts = jnp.diff(times)  # (ntimes-1,)
 
-        # === combine propagators to get the propagator at each requested time
-        def _reduce(prev_prop: Array, next_prop: Array) -> Array:
-            # notice the ordering of prev_prop and next_prop, want
-            # next_prop to be to the left of prev_prop
-            total_prop = next_prop @ prev_prop
-            return total_prop, total_prop
+        # === batch-compute the propagators on each time interval
+        Hs = jax.vmap(self.H)(times[:-1])  # (ntimes-1, n, n)
+        step_propagators = expm(-1j * delta_ts[:, None, None] * Hs)  # (ntimes-1, n, n)
```

Comment:
I think this is not quite right and will fail when Hs is batched. I confirmed this by changing the batch dimensions for H in the tests from (3,) and (4,) to (4,) and (4, 5) (unfortunate collision between batch dimension and tsave dimension in the tests) I would suggest instead
```suggestion
        dims = tuple(range(-Hs.ndim + 1, 0))
        delta_ts = jnp.expand_dims(delta_ts, dims)
        step_propagators = expm(-1j * delta_ts * Hs)  # (ntimes-1, nH, n, n)
```

Reply:
It goes without saying we should change the batch dims in the tests!

Reply:
Because the code is vectorized in `_vectorized_sepropagator` we don't need to care about batch dimensions for `H`, so we can write the code for a single `H` of shape `(n, n)` and the magic of vmap will do the rest. It's also the reason why I don't indicate the batching in the shape comments.

I edited the test with your suggestion to convince you. 😈

Reply:
Ahha! You are right!